lamagraph-core/src/Core/Concrete/ReduceRulesLambda.hs

```diff
@@ -0,0 +1,80 @@
+{- | All modules from Concrete are
+
+1. Compiled from some dsl in separate file
+
+2. Compiled from embedded dsl in `INet.Net` module
+
+3. Written by hands for tests
+-}
+module Core.Concrete.ReduceRulesLambda where
+
+import Clash.Prelude
+import Control.Lens hiding (Index, imap, (:>))
+import Core.MemoryManager.NodeChanges
+import Core.Node
+import Core.Reducer
+import Data.Maybe
+import INet.Net
+
+{- | Reduce rule for `Apply` and `Abs`
+
+<<docs/apply_to_lambda_rule.svg>>
+-}
+applyToLambdaRule ::
+  Vec 2 (Maybe AddressNumber) ->
+  LoadedNode 2 ->
+  LoadedNode 2 ->
+  ReduceRuleResult 2 2 2
+applyToLambdaRule _ loadedNodeLeft loadedNodeRight =
+  let arisingNodes = def
+      portsToEdgeEnds node = map (fromMaybe (error "Agent type of the node is incorrect")) (node ^. containedNode . secondaryPorts)
+      lE = portsToEdgeEnds loadedNodeLeft
+      rE = portsToEdgeEnds loadedNodeRight
+      arisingEdges = zipWith (\l r -> Just $ Edge l r) lE (reverse rE)
+   in ReduceRuleResult arisingEdges arisingNodes
+
+{- | Reduce rule fot `Erase` and `Abstract` or `Apply`
+
+<<docs/eps_apply_rule.svg>>
+
+__Note__ this rule is not symmetric!! Erasure node is first
+-}
+eraseToAbstractOrApplyRule ::
+  Vec 2 (Maybe AddressNumber) ->
```

Comment:
Why `Maybe AddressNumber`. All checks should be outside the logic.

Reply:
Without `Maybe` we need to change `Vec` size dynamically. It is impossible or very difficult. Or we should allocate a constant amount of memory, even if we don't need it. It is strange